Quickstart

lobrien opened the pull request:
DO NOT MERGE

morelen17 committed:
Fix "DISABLE_NGINX = 1" to "DISABLE_NGINX = true" (#8147)

Signed-off-by: Lenar Gabdrakhmanov <11724878+morelen17@users.noreply.github.com>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

harupy committed:
Remove TestDatabricksArtifactRepository (#8152)

Signed-off-by: harupy <hkawamura0130@gmail.com>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
First hack at restructuring quickstart

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Removed incorrect para

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Fixed indent / erroneous erro

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Comments and tweaks

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

mlflow-automation commented:
<!-- documentation preview -->


Documentation preview for df3dddbb4a95e234849d0e9360d69ad69a9c5c49 will be available [here](https://output.circle-artifacts.com/output/job/9a1eab33-c0d4-42bf-bd36-78a702f9a64d/artifacts/0/docs/build/html/index.html) when [this CircleCI job](https://circleci.com/gh/mlflow/mlflow/45151) completes successfully.

<details>
<summary>More info</summary>

- Ignore this comment if this PR does not change the documentation.
- It takes a few minutes for the preview to be available.
- The preview is updated when a new commit is pushed to this PR.
- This comment was created by https://github.com/mlflow/mlflow/actions/runs/4611314996.

</details>

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+How do I install MLFlow? 
+------------------------
+
+You may install MLflow from PyPi using pip:
```
We probably want to mention that there's an R and Java library, but we don't need to show the examples in R or Java.

lobrien replied:
Done in intro paragraphs and drill down.

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+
+And then navigate to http://localhost:5000 in your browser. You will see a page similar to:
+
+.. image:: _static/images/quickstart-ui-screenshot.png
+    :width: 800px
+    :align: center
```
If possible, would be great to grab a screenshot of the new UI here

lobrien replied:
New screenshot

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+    :width: 800px
+    :align: center
+
+You are in the **Default** experiment, which now contains the tracking data for your run. You can view the parameters, metrics, and artifacts of your run by clicking on the run name. 
```
Maybe a short sentence following this explaining an experiment is a collection of runs. Can place it between 1 and 2

lobrien replied:
Off-topic technical question: Is it best for an experiment to be a collection of runs with *identical* code (except for configuration values) or is it just "trying to get such-and-such a prediction from the domain, from day 1's terrible architecture to the super-bespoke architecture of day 1000"?

dbczumar replied:
Great question! Definitely the latter - `"trying to get such-and-such a prediction from the domain, from day 1's terrible architecture to the super-bespoke architecture of day 1000"?`?

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+How do I share my MLFlow runs and experiments?
+----------------------------------------------
+
+By default, MLFlow stores tracking data and artifacts in a ``./mlruns`` subdirectory of where you ran the code. You can change this behavior by setting the ``MLFLOW_TRACKING_URI`` environment variable to a different location. For instance, you can set it to a shared filesystem, a SQLAlchemy-compatible database, a tracking server, or a Databricks workspace.
```
Typically, to change the location, you would use `mlflow.set_tracking_uri()` from Python. The environment variable is helpful for provisioning a shared image or environment where data scientists (DS) can write their code without reasoning about the tracking URI.

lobrien replied:
Rewrote this H2 to be API-centric

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+
+By default, MLFlow stores tracking data and artifacts in a ``./mlruns`` subdirectory of where you ran the code. You can change this behavior by setting the ``MLFLOW_TRACKING_URI`` environment variable to a different location. For instance, you can set it to a shared filesystem, a SQLAlchemy-compatible database, a tracking server, or a Databricks workspace.
+
+You can run a tracking server on a network-accessible server by running:
```
Would be great to explain what a tracking server is and how it differs from the storage in `./mlruns`.

lobrien replied:
Added this in the first para of the "sharing" H2

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+How do I share my MLFlow runs and experiments?
+----------------------------------------------
+
+By default, MLFlow stores tracking data and artifacts in a ``./mlruns`` subdirectory of where you ran the code. You can change this behavior by setting the ``MLFLOW_TRACKING_URI`` environment variable to a different location. For instance, you can set it to a shared filesystem, a SQLAlchemy-compatible database, a tracking server, or a Databricks workspace.
```
Would also be super helpful when we say that the data is stored in `./mlruns` by default to explain what `mlflow ui` is doing. Basically, it visualizes the contents of that local filesystem directory.

lobrien replied:
Done

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+
+Now, when you run your code, it will send tracking data to the tracking server. To view the results, run the tracking UI on your local machine and point it to the tracking server:
+
+.. code-section::
+
+    .. code-block:: shell
+
+        mlflow ui --backend-store-uri http://192.168.0.1:5000
```
Should just be able to navigate to `http://192.168.0.1:5000` to view the UI

lobrien replied:
Removed this code and changed text to "browse to tracking server"

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+
+    If not:
+
+How do I use MLFlow with a Databricks workspace?
+------------------------------------------------
+
+tk 
```
Would be great to link to the Databricks docs for this; @dbczumar to find the relevant docs link.

lobrien replied:
I link to the Databricks QS.

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+tk 
+
+
+How do I store a model in MLFlow?
```
Autologging fortunately already stored, but would be great to indicate that mlflow.{flavor_name}.log_model() can be used to log models.

lobrien replied:
Wrote H2 and sample

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+
+tk 
+
+
```
Would be awesome to cover the model selection and production aspects. Show the user how to use the UI to pick a model, register with the MLflow Model Registry, and then deploy with `mlflow models serve` (local serving). Should link to the full MLflow Models & Model Registry references for more info about moving to production / production lifecycle.

lobrien replied:
I am worried about scope creep. I think there's a logical follow-on doc that's a quickstart for the MLOps-y review, validate, serve, monitor stuff, but I don't want to make the follow-along time for this page much >20 minutes.

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+
+For more options, see tk.
+
+How do I add MLFlow tracking to my Python code?
```
Would also be super helpful to diagram or introduce the workflow that we'll go through in the quickstart. Data -> Model training -> Model selection & registration -> Deployment

dbczumar replied:
Let's explicitly call out that the Data -> Training -> Registration is the data science workflow. Deployment is of interest to ML engineers / MLOps.

lobrien replied:
Added para in intro, plus diagram

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+- MLFlow example directory
+
+.. 
+    Drilldown for each H2. These could also be at the bottom
```
Would be great to include the drilldown in the H2s of the relevant sections

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+
+tk
+
+How do I do a hyperparameter sweep with MLFlow Tracking?
```
Can probably leave this out for now, but it would be great to call out Projects and Recipes briefly at the end of this workflow. "For more information about creating reproducible model training workflows, check out MLflow Projects.", "For prepackaged training and deployment solutions for classification and regression modeling, check out MLflow Recipes"

lobrien replied:
Removed this H2.

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+{>> Important things in original we don't want to lose:
+
+- mlflow skinny and extras and why you might choose one or t'other
+- Do we document / walkthrough virtualenvs or, e.g., conda or is that all assumed?
```
We should use the following for model serving because it doesn't depend on conda or virtualenv.

`mlflow models serve --env-manager local` 

Can then say that there's a way to restore the model's environment for serving and link to the MLflow Models page. Don't need to go into detail about conda / virtualenv here.

lobrien replied:
Punted this over to the drill down page

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+
+- mlflow skinny and extras and why you might choose one or t'other
+- Do we document / walkthrough virtualenvs or, e.g., conda or is that all assumed?
+- What about R? 
```
Just calling out that the library can be installed in Java / R is plenty here. R isn't widely used.

lobrien replied:
Done in the drill down. I actually ended up linking over to the R API page, since there's an intro page. For Java, I "due to the complexity of the Java ecosystem..." bailed.

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+- mlflow skinny and extras and why you might choose one or t'other
+- Do we document / walkthrough virtualenvs or, e.g., conda or is that all assumed?
+- What about R? 
+- Mac python on install (relates to env answer)
```
Keeping the mac note seems helpful, since macs are pretty ubiquitous in industry at this point

lobrien replied:
Done in the drill down

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+- What about R? 
+- Mac python on install (relates to env answer)
+- repo contains a big helpful directory of examples 
+- WORKER_TIMEOUT issue 
```
Don't have to worry about this for quickstart

lobrien replied:
Deleted. Haven't re-added in the drill down

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
 Quickstart
 ==========
 
+Install MLFlow, instrument your code, and view a run's results in a few minutes:
```
```suggestion
Install MLflow, instrument your code, and view a run's results in a few minutes:
```

lobrien commented:
@BenWilson2

Commit:
Incorporating GH review comments

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Use w Databricks ws

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Added save and load models

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Updated images

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Renamed mlflow images back to quickstart

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Renamed screenshot for consistency

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Removed TODOs and tks

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Title/subtile formatting

In Sphinx, titles are === and subtitles are —.

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Improved link

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Reversed H2 styles

Back to === for H1, — for H2. Apparently it’s all good, so long as you’re consistent. So what do we use for H3s?

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

BenWilson2 commented on docs/source/quickstart.rst:
```diff
@@ -77,12 +89,12 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
         if __name__ == "__main__":
             # Log a parameter (key-value pair)
-            log_param("param1", randint(0, 100))
+            log_param("config_value", randint(0, 100))
```
it might be nice to show the dict-based logging here as well

```
# Log a collection of parameters
run_params = {
  "optimizer": "sgd",
  "seed": 98765,
  "n_iterations": 2500,
}
mlflow.log_params(run_params)
```

lobrien replied:
TODO

BenWilson2 commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +103,193 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
-Launch a Tracking Server on a Remote Machine
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-:ref:`Launch a tracking server <tracking_server>` on a remote machine.
+- An **MLModel** file in YAML format specifying the model's **flavor** (or **flavors**);
```
```suggestion
- An **MLModel** file in YAML format specifying the model's **flavor** (or **flavors**), dependencies, signature (if supplied), and important metadata;
```

lobrien replied:
Added

BenWilson2 commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +103,193 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
-:ref:`Launch a tracking server <tracking_server>` on a remote machine.
+- An **MLModel** file in YAML format specifying the model's **flavor** (or **flavors**);
+- The various files required by the model's flavor(s) to instantiate the model. This will often be a serialized Python object; and
+- Files necessary for recreating the model's runtime environment (for instance, a **conda.yaml** file)
```
Would it be helpful to show a screenshot of the contents of an MLModel file and the supplementary saved objects? 
![Screenshot 2023-04-12 at 9 48 49 AM](https://user-images.githubusercontent.com/39283302/231478456-27873965-80a7-47f3-877d-f5ebd688e121.png)

lobrien replied:
I think this should be in the drill down coverage

lobrien replied:
I added the "input example if specified"

BenWilson2 commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +103,193 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
+Note that while ``log_model`` saves environment-specifying files such as **conda.yaml** and **requirements.txt**, ``load_model`` does not automatically recreate that environment. To do so, you need to use your preferred method (**conda**, **virtualenv**, **pip**, etc.), using the artifacts saved by ``log_model``.
+
+To learn more about loading runs, see :ref:`quickstart_drilldown_log_and_load_model`.
+
```
It might be nice to show the programmatic access pattern for retrieving the model_uri via the run context:

```
with mlflow.start_run() as run:
  mlflow.sklearn.log_model(
    my_model,
    artifact_path="model",
  )


model_uri = f"runs:/{run.info.run_id}/model"
loaded_model = mlflow.sklearn.load_model(model_uri)
```

lobrien replied:
TODO

BenWilson2 commented on docs/source/quickstart_drilldown.rst:
```diff
@@ -0,0 +1,144 @@
+---------------------------------
+
+tk
+
```
examples of how to use the search functionality :) There are a lot of issues filed with users being confused by the syntax / utility of it

lobrien replied:
TODO

lobrien commented on docs/source/quickstart.rst:
```diff
@@ -3,6 +3,219 @@
+     - Values updated during the run (for instance, accuracy)
+     - ``mlflow.log_metric``
+   * - Artifacts
+     - Files produced by the run (for instance, model weights)
+     - ``mlflow.log_artifacts``
```
Done.

lobrien commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +103,193 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
 
-        # Log a parameter (key-value pair)
-        mlflow_log_param("param1", 5)
+Once you've run your code, you may view the results with MLflow's tracking UI. To start the UI, run:
```
I've done this in the opening sentences of the "sharing" H2. Not super-happy with the flow of the text, but it covers the bases.

lobrien commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +103,193 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
 
-    .. code-block:: R
+- For more details on autolog, including a list of what libraries are supported, see :ref:`quickstart_drilldown_autolog`.
+- For additional functions such as ``log_text`` and ``log_image``, see :ref:`quickstart_drilldown_tracking_api`.
```
Done.

Commit:
Incorporate PR comments

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
mv images to subdir per style

Ref. https://databricks.atlassian.net/wiki/spaces/UN/pages/2889122429/Images+graphics+diagrams+and+screenshots+in+documentation#Where-do-I-store-images%3F

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Create image-sources dir

Ala. https://databricks.atlassian.net/wiki/spaces/UN/pages/2889122429/Images+graphics+diagrams+and+screenshots+in+documentation#Where-do-I-store-images%3F

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Modifying cross-refs to point to existing docs

When possible, point to existing docs.

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Finalizing cross-references

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien commented:
## Related Issues/PRs

None that I could see. I searched **quickstart** in issues and none seem relevant to this text. 

## What changes are proposed in this pull request?

Rewrite of the MLFlow quickstart for focus, terseness, and SEO. 

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [X] Manual tests (describe details, including test results, below)

Local builds and review. Tech review from engineers, editorial peer-review from Carolyn G.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

New quickstart focused on time-to-live for Python data scientists.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

(Has an H2 on using MLFlow tracking with a Databricks workspace)

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [X] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

lobrien commented:
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>
Authored-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien commented:
The force-push was to sign all my previous commits.

Commit:
Cross-references

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Quickstart drilldown

Hid in TOC and edited content

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

carolyn-gronlund reviewed (changes requested):
Looks good. Some changes requested. :>)

lobrien committed:
Update docs/source/tracking.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/tracking.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart_drilldown.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart_drilldown.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart_drilldown.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart_drilldown.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart_drilldown.rst

Co-authored-by: CJ Gronlund <103294599+carolyn-gronlund@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Formatting tweaks

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien commented:
I still get 2 warnings on the build, but they relate to references from the docstrings. In particular, I tried to chase down the warning "mlflow/mlflow/sklearn/__init__.py:docstring of mlflow.sklearn.autolog:130:" by looking at the docstring for autolog, but I didn't find any :ref: references. So hopefully those aren't blocking warnings.

lobrien commented on docs/source/quickstart_drilldown.rst:
```diff
@@ -0,0 +1,133 @@
+When you run the example, it outputs an MLflow run ID for that experiment. If you look at
+``mlflow ui``, you will also see that the run saved a ``model`` folder containing an ``MLmodel``
+description file and a pickled scikit-learn model. You can pass the run ID and the path of the model
+within the artifacts directory (here "model") to various tools. For example, MLflow includes a
```
DONE

Commit:
Formatting

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien commented:
Force-push to sign 2 commits

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -1,88 +1,110 @@
 -----------------
 
-You install MLflow by running:
+Install MLflow from PyPi using pip:
```
```suggestion
Install MLflow from PyPI using pip:
```

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -1,88 +1,110 @@
+- (Optional) Run a tracking server to share results with others
+- (Optional) Use Databricks to store your results
+- Store the models produced by your runs
+- Run a model from a previous run
```
Nit: We typically refer to "model loading" and "model inference", rather than "running a model"

```suggestion
- Load a model from a previous run for inference
```

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -1,88 +1,110 @@
+- Store the models produced by your runs
+- Run a model from a previous run
+
+As a data scientist, your explorations involve running your evolving training code many times. MLflow Tracking allows you to record important information your run, review and compare it with other runs, and share results with others. As an ML Engineer or MLOps professional, it allows you to compare, share, and deploy the best models produced by the team.
```
I love that we call out `As an ML Engineer or MLOps professional, it allows you to compare, share, and deploy the best models produced by the team.` at the end, but the quickstart doesn't cover these topics. We should definitely follow up with quickstart content for ML Engineers / MLOps. Just leaving this here as a note, no immediate action item

Commit:
Fixed code per blacken-docs failure

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,206 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
+And then navigate to http://localhost:5000 in your browser. You will see a page similar to:
 
-        mlflow_ui()
+.. image:: _static/images/quickstart/quickstart_ui_screenshot.png
```
Thanks for updating the screenshot! A couple small asks:

1. Can we show some metrics and parameters in the "Table view" screenshot? Right now, it's hard to see that MLflow recorded these.

2. Can we include a second screenshot showcasing the "Chart view" with a couple of charts?

lobrien replied:
For both table and chart, the top-level screen shows the runs, not the metrics/params.

So I need to drill down, which is fine, but: 

For the RandomForestRegressor sample, it does it all in a single pass, so the metrics don't go up over time. So I need to use a screenshot from a different sample, which is fine, but I'll need to write around it a little.

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: Corey Zumar <39497902+dbczumar@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,206 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
-You can easily run existing projects with the ``mlflow run`` command, which runs a project from
-either a local directory or a GitHub URI:
+- Calling ``mlflow.set_tracking_uri`` in your code; or
+- Setting the ``MLFLOW_TRACKING_URI`` environment variable
 
-.. code-block:: bash
+With either option, you can track your runs in a shared filesystem, a SQLAlchemy-compatible database, a tracking server, or a Databricks workspace.
```
Love that we're calling this out! I found it a bit confusing that we're telling people to call this Python API / set this environment variable before we tell them about the possible values they can use (file system URI for shared filesystem, SQLAlchemy URI for database, HTTP/S URL for tracking server, etc).

Perhaps we should move this below the discussion of "what is a tracking server?". It would also be great to include a couple of examples showing how you can use `set_tracking_uri` and `MLFLOW_TRACKING_URI` with a shared filesystem and with a SQL database. Can we also link to the relevant MLflow Tracking docs so that users can find more info?

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,206 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
+- An **MLModel** file in YAML format specifying the model's **flavor** (or **flavors**), dependencies, signature (if supplied), and important metadata;
+- The various files required by the model's flavor(s) to instantiate the model. This will often be a serialized Python object;
+- Files necessary for recreating the model's runtime environment (for instance, a **conda.yaml** file); and
+- Optionally, an input sample
```
Nit: "Input example" is used throughout the docs and in the API

```suggestion
- Optionally, an input example
```

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: Corey Zumar <39497902+dbczumar@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,206 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
+- Optionally, an input sample
 
-  .. code-section::
+When using autologging, MLflow will automatically log the run's model. You can also log a model manually by calling ``mlflow.{library_module_name}.log_model``. In addition, if you wish to load the model soon, it may be convenient to output the run's ID directly to the console. For that, you'll need the object of type ``mlflow.ActiveRun`` for the current run. You get that object by wrapping all of ...
```
Nit: I think it's a bit unclear what "run's model" means; a run could have multiple models or no models at all

```suggestion
When using autologging, MLflow will automatically log models to the runs that it creates. You can also log a model manually by calling ``mlflow.{library_module_name}.log_model``. In addition, if you wish to load the model soon, it may be convenient to output the run's ID directly to the console. For that, you'll need the object of type ``mlflow.ActiveRun`` for the current run. You get that object by wrapping all of your logging code in a ``with mlflow.start_run() as run:`` block. For example:
```

dbczumar replied:
Nit: Can we link to the Python API docs for `start_run()` and `ActiveRun`?

lobrien replied:
DONE

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: Corey Zumar <39497902+dbczumar@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart.rst

Co-authored-by: Corey Zumar <39497902+dbczumar@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,206 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
+        predictions = model.predict(X_test)
+        print(predictions)
+
+Note that while ``log_model`` saves environment-specifying files such as **conda.yaml** and **requirements.txt**, ``load_model`` does not automatically recreate that environment. To do so, you need to use your preferred method (**conda**, **virtualenv**, **pip**, etc.), using the artifacts saved by ``log_model``.
```
Great call-out! Can we also state that MLflow makes this easy to do? Wouldn't want users to worry that they have to set up the conda / virtualenv by themselves

lobrien replied:
DONE

dbczumar commented on docs/source/quickstart_drilldown.rst:
```diff
@@ -0,0 +1,135 @@
+     - Lightweight MLflow package without SQL storage, server, UI, or data science dependencies.
+   * - mlflow[extras]
+     - ``pip install mlflow[extras]``
+     - MLflow package with all dependencies needed to run various MLflow flavors. These dependencies are listed in doc:`https://github.com/mlflow/mlflow/blob/master/requirements/extra-ml-requirements.txt`
```
Did we mean for this to say doc:`https://github.com/mlflow/mlflow/blob/master/requirements/extra-ml-requirements.txt`?

I think it would be better to make this a link to the file and remove the text "doc:"

lobrien replied:
lol. DONE

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -1,88 +1,110 @@
 
-Installing MLflow
+
+In 15-20 minutes, you will:
```
15-20 minutes sounds long. I think users can get through this in 10 minutes. How do you feel about saying "10 minutes"?

lobrien replied:
Compromise "less than 15 minutes"? :-)

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,206 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
 
-    By default the server runs on port 5000. If that port is already in use, use the `--port` option to
-    specify a different port. For example: ``mlflow models serve -m runs:/<RUN_ID>/model --port 1234``
+Use MLflow with a Databricks workspace
```
I think we should make this a subheader of "Share MLflow Runs and experiments"

lobrien replied:
~~If we do this, it will drop off the TOC. ~~ DONE

BenWilson2 commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,206 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
+
+For more information, including a list of supported model flavors and storing your own flavor, see :ref:`models_built-in-model-flavors`.
+
+Load a model from a specific training run for inference
+--------------------------------------------
```
```suggestion
Load a model from a specific training run for inference
------------------------------------------------------
```
.rst linting is complaining about this. We have lint checks enabled that mandate the section delimiter must be at least as long as the title.

BenWilson2 commented on docs/source/quickstart_drilldown.rst:
```diff
@@ -0,0 +1,135 @@
+- PyPI (``pip install mlflow``)
+- conda-forge (``conda install -c conda-forge mlflow``)
+- CRAN (``install.packages("mlflow")``)
+- Maven Central (``mlflow-client``, ``mlflow-parent``, ``mlflow-scoring``, ``mlflow-spark``)
```
Can we move the CRAN / Maven Central to the `R and Java` section below and then add a `Python` section for the `PyPI` and `conda-forge` elements here?

lobrien replied:
DONE

BenWilson2 commented on docs/source/quickstart_drilldown.rst:
```diff
@@ -0,0 +1,135 @@
+
+We strongly recommend using a virtual environment manager on Macs. We always recommend using virtual environments, but they are especially important on Mac OS X because the system ``python`` version varies depending on the installation and whether you've installed the Xcode command line tools.
+
+Virtual environment managers
```
Is the intention with this title to allude to conda? The only thing listed below that is an environment manager is conda (which interfaces with the conda-forge repo), but pip is just a package dependency resolver (virtualenv would be the 'standard python' environment manager). 
CRAN is similar to pypi; more of a dependency management tool. 
Maven is definitely not an environment manager. jEnv is, but that's entirely up to users to figure out how to make that work.

lobrien replied:
DONE

BenWilson2 commented on docs/source/quickstart_drilldown.rst:
```diff
@@ -0,0 +1,135 @@
+serve it. MLflow also includes tools for running such models locally and exporting them to Docker
+containers or commercial serving platforms.
+
+To illustrate this functionality, the ``mlflow.sklearn`` package can log scikit-learn models as
```
```suggestion
To illustrate this functionality, the ``mlflow.sklearn`` flavor can log scikit-learn models as
```

BenWilson2 commented on docs/source/quickstart_drilldown.rst:
```diff
@@ -0,0 +1,135 @@
+
+To illustrate this functionality, the ``mlflow.sklearn`` package can log scikit-learn models as
+MLflow artifacts and then load them again for serving. There is an example training application in
+``sklearn_logistic_regression/train.py`` that you can run as follows:
```
Can we change this to a link on github master to this example in case someone wants to directly see the code here?

lobrien replied:
DONE

BenWilson2 commented on docs/source/quickstart_drilldown.rst:
```diff
@@ -0,0 +1,135 @@
+
+.. code-block:: bash
+
+    python sklearn_logistic_regression/train.py
```
Can we tell users that to run this from the repository root that they will have to `cd examples` or update the snippet to execute the path to this example from the repository root `python examples/sklearn_logistic_regression/train.py`?

lobrien replied:
DONE. (the latter.)

BenWilson2 commented on docs/source/quickstart_drilldown.rst:
```diff
@@ -0,0 +1,135 @@
+   * - mlflow-skinny
+     - ``pip install mlflow-skinny``
+     - Lightweight MLflow package without SQL storage, server, UI, or data science dependencies.
+   * - mlflow[extras]
```
Do we want to include the 'install dev' version here as well? We end up telling users in reported issues to do this quite frequently to apply a merged-but-not-released hotfix to master quite often.

BenWilson2 commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,206 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
+- :ref:`MLflow tutorials and examples <tutorials-and-examples>`
+- Use the MLflow Registry to store and share versioned models, see :ref:`registry`
+- Use MLflow Projects for packaging your code in a reproducible and reusable way, see :ref:`projects`
+- Use MLflow Recipes (experimental) to create workflows for faster iterations and easier deployment, see :ref:`recipes`
```
We can remove "experimental" references for MLflow Recipes. It's been stabilized with the 2.3.0 release

lobrien replied:
DONE

Commit:
Added new graphics to mlflow ui disc

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart_drilldown.rst

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien committed:
Update docs/source/quickstart_drilldown.rst

Co-authored-by: Ben Wilson <39283302+BenWilson2@users.noreply.github.com>
Signed-off-by: Larry O'Brien <lobrien@knowing.net>
Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Responding to GH comments

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

lobrien commented on docs/source/quickstart_drilldown.rst:
```diff
@@ -0,0 +1,135 @@
+
+For more information, see :doc:`models`.
+
+Run MLflow projects
```
Hmm... This is an artifact of the previous quickstart. It seems valuable, but I think the proper final place for this is in the to-be-written QS for MLOps. Hmm... At the moment, it's really just a valuable piece of writing that's orphaned. Hmmm... I think I'm going to cut this and keep the fragment in my notes.

lobrien commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,206 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
-    By default the server runs on port 5000. If that port is already in use, use the `--port` option to
-    specify a different port. For example: ``mlflow models serve -m runs:/<RUN_ID>/model --port 1234``
+Use MLflow with a Databricks workspace
+----------------------------------------
```
DONE

Commit:
GH comments

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

Commit:
Merge branch 'main' into quickstart

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>

# Conflicts:
#	tests/store/artifact/test_databricks_artifact_repo.py

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -1,88 +1,110 @@
+
+- Install MLflow
+- Add MLflow tracking to your code
+- View run metrics in the MLflow tracking UI
```
```suggestion
- View runs and experiments in the MLflow tracking UI
```

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,227 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-    curl -d '{"dataframe_split": {"columns": ["x"], "data": [[1], [-1]]}}' -H 'Content-Type: application/json' -X POST localhost:5000/invocations
+You need to configure Mlflow to use your Databricks workspace (To get started with Databricks, see: `Get started: Account and Workspace setup <https://docs.databricks.com/getting-started/index.html>`_). You will need to know the URL of your Databricks workspace. You can find the URL in the Configuration page of the workspace:
```
```suggestion
You need to configure MLflow to use your Databricks workspace (To get started with Databricks, see: `Get started: Account and Workspace setup <https://docs.databricks.com/getting-started/index.html>`_). You will need to know the URL of your Databricks workspace. You can find the URL in the Configuration page of the workspace:
```

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,227 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
+        db = load_diabetes()
+        X_train, X_test, y_train, y_test = train_test_split(db.data, db.target)
+
+        model = mlflow.sklearn.load_model("runs:/97fd7ade5106ee341e0b4c63a53a9776231")
```
Can we use the run ID from the screenshot above for consistency?

dbczumar commented on docs/source/quickstart.rst:
```diff
@@ -91,187 +113,227 @@ as follows (this example is also included in ``quickstart/mlflow_tracking.py``):
+        predictions = model.predict(X_test)
+        print(predictions)
+
+Note that while ``log_model`` saves environment-specifying files such as **conda.yaml** and **requirements.txt**, ``load_model`` does not automatically recreate that environment. To do so, you need to use your preferred method (**conda**, **virtualenv**, **pip**, etc.), using the artifacts saved by ``log_model``. If you serve your model with ``mlflow models serve`` or ``mlflow run``, MLflow will  ...
```
Nit: `mlflow run` isn't used for model serving. 

```suggestion
Note that while ``log_model`` saves environment-specifying files such as **conda.yaml** and **requirements.txt**, ``load_model`` does not automatically recreate that environment. To do so, you need to use your preferred method (**conda**, **virtualenv**, **pip**, etc.), using the artifacts saved by ``log_model``. If you serve your model with ``mlflow models serve``, MLflow will automatically recreate the environment. Those commands also accept an ``--env-manager`` option for even more control. (This is described in detail in :ref:`model-enviroment-management`.)
```

Can we also mention that ``mlflow.pyfunc.spark_udf()`` supports an `env_manager` flag to recreate the model environment during batch inference with Spark? It's set to ``local`` by default but can be set to ``virtualenv`` and ``conda``.

dbczumar reviewed (approved):
Amazing work, @lobrien! Thank you so much! Approved pending resolution of small comments. :D

Commit:
Final final? :-)

Signed-off-by: Larry O’Brien <larry.obrien@databricks.com>
